data_engineering.qmd

```diff
@@ -1,121 +1,526 @@
+
+Pivacy and confidentiality is also a big issue. Datasets containing sensitive or personal information pose privacy concerns when shared or used. Synthetic data, being artificially generated, doesn’t have these direct ties to real individuals, allowing for safer use while preserving essential statistical properties.
+
+Generating synthetic data, especially once the generation mechanisms have been established, can be a more cost-effective alternative. In the aforementioned security application scenario, synthetic data eliminates the need for breaking multiple windows to gather relevant data.
+
+Many embedded use-cases deal with unique situations, such as manufacturing plants, that are difficult to simulate. Synthetic data allows researchers complete control over the data generation process, enabling the creation of specific scenarios or conditions that are challenging to capture in real life.
+
+While synthetic data offers numerous advantages, it is essential to use it judiciously. Care must be taken to ensure that the generated data accurately represents the underlying real-world distributions and does not introduce unintended biases.
 
-- Data Sources: crowdsourcing, pre-existing datasets etc.
-- Data Types: Structured, Semi-Structured, and Unstructured
-- Real-time Data Processing in Embedded Systems
 
 ## Data Storage
+Data sourcing and data storage go hand-in-hand and it is necessary to store data in a format that facilitates easy access and processing. Depending on the use case, there are various kinds of data storage systems that can be used to store your datasets.
+
+  ----------------------------------------------------------------------------
+                 **Database**        **Data Warehouse**    **Data Lake**
+  -------------- ------------------- --------------------- -------------------
+  **Purpose**    Operational and     Analytical            Analytical
+                 transactional                             
+
+  **Data type**  Structured          Structured            Structured,
+                                                           semi-structured
+                                                           and/or unstructured
+
+  **Scale**      Small to large      Large volumes of      Large volumes of
+                 volumes of data     integrated data       diverse data
+
+  **Examples**   MySQL               Google BigQuery,      Google Cloud
+                                     Amazon Redshift,      Storage, AWS S3,
+                                     Microsoft Azure       Azure Data Lake
+                                     Synapse.              Storage
+  ----------------------------------------------------------------------------
+
```

Comment:
I found this section about Data Storage to be really interesting and something I've never really interacted with before! While this table is super informative already, it would have been great if there was a more in depth analysis of these systems and what people generally use in practice. In particular, I imagine that data is never really stored on tiny devices themselves, and so it must be sent from data sources on other devices. Does communication and retrieval differ for each of these systems and does that impact their integration into TinyML pipelines? Is the biggest factor in picking between these systems the cost?

Reply:
Agreed, it would be great to have the similar discussion of implications for TinyML like you did for the Data Sourcing section.